src/lib/rng/stateful_rng/stateful_rng.cpp

```diff
@@ -34,83 +34,63 @@ bool Stateful_RNG::is_seeded() const
    return m_reseed_counter > 0;
    }
 
-void Stateful_RNG::add_entropy(const uint8_t input[], size_t input_len)
-   {
-   lock_guard_type<recursive_mutex_type> lock(m_mutex);
-
-   update(input, input_len);
-
-   if(8*input_len >= security_level())
-      {
-      reset_reseed_counter();
-      }
-   }
-
-void Stateful_RNG::initialize_with(const uint8_t input[], size_t len)
+void Stateful_RNG::initialize_with(std::span<const uint8_t> input)
    {
    lock_guard_type<recursive_mutex_type> lock(m_mutex);
 
    clear();
-   add_entropy(input, len);
+   add_entropy(input);
    }
 
-void Stateful_RNG::randomize(uint8_t output[], size_t output_len)
+void Stateful_RNG::generate_batched_output(std::span<uint8_t> output, std::span<const uint8_t> input)
    {
-   randomize_with_input(output, output_len, nullptr, 0);
-   }
-
-void Stateful_RNG::randomize_with_ts_input(uint8_t output[], size_t output_len)
-   {
-   uint8_t additional_input[20] = { 0 };
-
-   store_le(OS::get_high_resolution_clock(), additional_input);
-
-#if defined(BOTAN_HAS_SYSTEM_RNG)
-   System_RNG system_rng;
-   system_rng.randomize(additional_input + 8, sizeof(additional_input) - 8);
-#else
-   store_le(OS::get_system_timestamp_ns(), additional_input + 8);
-   store_le(OS::get_process_id(), additional_input + 16);
-#endif
-
-   randomize_with_input(output, output_len, additional_input, sizeof(additional_input));
-   }
-
-void Stateful_RNG::randomize_with_input(uint8_t output[], size_t output_len,
-                                        const uint8_t input[], size_t input_len)
-   {
-   if(output_len == 0)
-      return;
-
-   lock_guard_type<recursive_mutex_type> lock(m_mutex);
+   BOTAN_ASSERT_NOMSG(!output.empty());
 
    const size_t max_per_request = max_number_of_bytes_per_request();
 
    if(max_per_request == 0) // no limit
       {
       reseed_check();
-      this->generate_output(output, output_len, input, input_len);
+      this->generate_output(output, input);
       }
    else
       {
-      while(output_len > 0)
+      size_t output_length = output.size();
+      size_t output_offset = 0;
+      while(output_length > 0)
          {
-         const size_t this_req = std::min(max_per_request, output_len);
-
-         /*
-         * We split the request into several requests to the underlying DRBG but
-         * pass the input to each invocation. It might be more sensible to only
-         * provide it for the first invocation, however between 2.0 and 2.15
-         * HMAC_DRBG always provided it for all requests so retain that here.
-         */
```

Comment:
We altered the behavior in the sense described here. I.e. using the input only with the first round. As a result `a` and `b` result in the same byte stream (assuming equally seeded RNGs),

```C++
auto a = rng.random_vec(64*1024) + rng.random_vec(64*1024) + rng.random_vec(64*1024);
auto b = rng.random_vec(3*64*1024)
```

See also: https://github.com/randombit/botan/pull/3195/commits/fff82489dabf6786734119ca3e74246a5169db10

@randombit Do you think that's okay?

Reply:
If we keep this we probably also want to mention it int the PR description.

Reply:
This seems OK to me. When I introduced `Stateful_RNG` I wanted to retain the existing behavior of `HMAC_DRBG`, but this new behavior actually does make more sense and if we are going to change it, 3.0 is a good time for it.

Reply:
Does this mandate a mention in the migration guide? It might break people's "deterministic" algorithms under certain circumstances after all.

Reply:
Yes definitely worth a mention. Hopefully people aren't relying on this behavior since it's not entirely standard, but Hyrum's Law applies. The change in behavior is at least easy to work around in application code, by breaking up calls that are too long and intentionally repeating the input for each call, mimicking what we did internally.